### PR TITLE
fix(sparse): correct SM100 blk64 output when seqlen_k % 64 != 0

### PR DESCRIPTION
**Pull request by hsr1234563**
When seqlen_k is not divisible by 64, the last KV block contains fewer than 64 valid tokens. The packed rank-6 TMA view rounded up to ceil_div(seqlen_k, 64) full blocks, causing TMA to read beyond the allocated buffer. Out-of-bounds K values (zero-filled) produced score=0, giving exp(0)=1 softmax weight and corrupting the output.

Two-part fix:

1. Auto-inject block_sizes[-1] = seqlen_k % 64 (primary correctness fix) When seqlen_k % 64 != 0 and the user did not provide block_sizes, generate a block_sizes tensor with 64 for full blocks and seqlen_k%64 for the partial tail. This triggers apply_block_size_mask_64 which sets out-of-bounds attention scores to -inf, giving them zero softmax weight. Handles the interaction with phantom-block masking (variable block nums) by checking user_provided_block_sizes rather than has_block_sizes so both corrections apply simultaneously.

2. Add use_exact_kv_layout kernel variant (TMA safety fix) Builds a 5D TMA descriptor with the true seqlen_k extent and Int32 strides so TMA zero-fills out-of-bounds positions instead of reading arbitrary memory beyond the allocation.

Also relax the wrapper N % C == 0 requirement for vsa_sm100_blk64 (other backends still require alignment) and switch NB to ceiling division so plan() accepts partial seqlen_k without raising ValueError. Sage FP8 continues to reject seqlen_k % 64 != 0 via validate_sm100_blk64_fp8_sage (D-major layout incompatible with partial tail masking).

Tests: add test_vsa_blk64_partial_kv_tail (seqlen_k=65/127/193) and test_vsa_blk64_partial_kv_tail_kv_splits (kv_splits=2/4) to cover the single-kernel and split-KV combine paths.

AI-assisted

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed block-sparse attention for key/value sequence lengths that are not multiples of 64.
  - Prevented out-of-bounds data from affecting softmax calculations, avoiding invalid outputs such as NaN or Inf.
  - Ensured correct results when processing attention with multiple KV splits.

- **Compatibility**
  - Added support for partial final KV blocks in the SM100 block-sparse attention backend.
  - Partial KV tails remain unsupported for Sage FP8 layouts and now produce a validation error.

### CODE REVIEW

File: flashinfer/sparse.py [hunk 1]
```diff
@@ -999,10 +1000,11 @@ def plan(
                 causal,
                 pos_encoding_mode,
                 logits_soft_cap,
+                require_n_aligned=False,  # blk64 supports partial last KV block
             )
 
             MB = M // R
-            NB = N // C
+            NB = (N + C - 1) // C
```
[the remark this answers is not in the document]
Comment (hsr1234563) [reply flashinfer/sparse.py:1009]: Done

**Comment by qsang-nv** [comment]
Non-blocking follow-ups on the contract, duplication, and test scope; the root cause and the two-part fix look right to me.

**1. The partial-tail constraint on `block_sizes` is neither documented nor validated.** The tail injection is gated on `not user_provided_block_sizes` while `use_exact_kv_layout` is not, so a caller supplying `block_sizes` now owns the tail entry: an entry larger than the physical tail leaves unmasked zero logits that receive non-zero softmax probability mass, with no error raised. The docstring ("Actual token count per KV block ... Pass None to skip per-block padding masking (assumes all blocks are full)") arguably already makes that invalid input, but it predates partial-tail support and no longer describes what `None` does either — so the requirement is unstated exactly where it starts to matter. Worth saying explicitly whether the tail entry is purely the caller's responsibility. `_vsa_run_core_blk64` always passes `None`, so only direct kernel-API callers are affected.

**2. Sage FP8 rejects the partial tail only at run time.** The check lives in `validate_sm100_blk64_fp8_sage`, but `plan()`'s `vsa_sm100_blk64` branch already has `is_sage_fp8` and `N`, so `plan(N=193, q_scale=...)` succeeds and only the first `run()` raises.

**3. Implementation and comment duplication.** The new `elif const_expr(self.use_exact_kv_layout)` block in `__call__` is identical to the `if const_expr(self.use_int64_kv_strides)` block except for `Int64` → `Int32` (~47 lines to keep in sync by hand), and "true `seqlen_k` extent → TMA zero-fills" is restated at five sites, "rank-6 rounds up → reads past the buffer" at three; ~42 of the ~125 non-test lines added are comments. If factoring the two branches through a constexpr-selected integer type is not supported cleanly by CuTe DSL, a small shared helper — or a concise comment explaining why the duplication is required — would still make the intent explicit.

**4. Test scope.** Both new tests go through the wrapper, which always passes `q2k_block_nums`, so `has_variable_block_nums=True` and `block_sizes` is already allocated by the pre-existing phantom-masking branch. The direct-API path with a uniform `block_sparse_num` is a different specialization (`has_variable_block_nums` is in the compile key) and the only one exercising the new `if block_sizes is None` allocation; it has no coverage. The new comparisons also round the FP32 reference to bf16 before `assert_close`, folding an extra rounding into the tolerance.
